Type `args` and `kwargs` in `pipe`

paw-lu opened the pull request:
- [x] Closes #738
- [x] Tests added: Please use `assert_type()` to assert the type of any return value

Follow up to https://github.com/pandas-dev/pandas/issues/56359. Type out `args` and `kwargs` in `pipe` methods, and make all instances of the `pipe` method consistent in their typing ( #738).

paw-lu commented:
Started out typing out frame's `pipe` to test things out, and already running into issues 😅.  Surprisingly to me, mypy  is complaining on the return type of the `func` used in `pipe.`

```sh
tests/test_frame.py:1393: error: Expression is of type <nothing>, not "DataFrame"  [assert-type]
tests/test_frame.py:1394: error: Argument 1 to "pipe" of "DataFrame" has incompatible type "Callable[[DataFrame, int, str], DataFrame]"; expected "Callable[[DataFrame, int, str], <nothing>] | tuple[Callable[..., <nothing>], str]"  [arg-type]
```

Looking into this!

paw-lu commented on tests/test_frame.py:
```diff
@@ -1386,6 +1386,16 @@ def baz(val: Styler) -> str:
+
+    check(
+        assert_type(
+            pd.DataFrame({"a": [1], "b": [1]}).pipe(qux, 1, y="a"), pd.DataFrame
```
> Isn't the problem here that y is not an argument of qux?

Embarrassing. Yeah this might be it—forgot to commit my entire diff. Thanks!

> a function with a "normal" argument, a keyword-only argument, and a positional-only argument

Agree on all of these. Didn't think of including the positional only argument, good call there.

> an invalid call by providing an unexpected keyword argument (will need ignores and be inside TYPE_CHECKING_INVALID_USAGE)

Yeah plus also feeding in the wrong type as well to positional, positional-only, keyword, etc.

Dr-Irv replied:
> It might be nice to have at least two tests:
> 
> * a function with a "normal" argument, a keyword-only argument, and a positional-only argument
> * an invalid call by providing an unexpected keyword argument (will need ignores and be inside `TYPE_CHECKING_INVALID_USAGE`)

I agree with @twoertwein  here.  Should also test the `tuple` kind of argument as well.

paw-lu replied:
> Should also test the tuple kind of argument as well.

Oh yeah, kinda assumed there were tests already for that since I didn't touch that version of the typing, but I'll look and if they don't already exists—I'll add them

Dr-Irv reviewed (changes requested):
There is an issue here with respect to having `pipe` defined in `core/frame.pyi` and `core/series.pyi` .  I think if you took the concept proposed in this PR and used it in `generic.pyi` and removed it from `core/frame.pyi`, things would work for both `Series` and `DataFrame`.

So can you make that change, and then have tests for both `Series.pipe()` and `DataFrame.pipe()` ?

paw-lu commented on pandas-stubs/core/frame.pyi:
```diff
@@ -1832,9 +1836,9 @@ class DataFrame(NDFrame, OpsMixin):
-        func: Callable[..., TType] | tuple[Callable[..., TType], _str],
-        *args,
-        **kwargs,
+        func: Callable[Concatenate[Self, P], TType] | tuple[Callable[..., TType], _str],
```
I wish. I don't think Python's typing is flexible enough to account for the `tuple` version of the arg.

You'd want to express that _some_ argument (no longer just the first) in `func` takes in `Self`, and the rest of the params can be passed in via `args` and `kwargs`. This goes way beyond what `Concatenate` can do.

If you happen to know how to type this though happy to implement it. I'm just leaving it as is because I can't think of a way to improve it.

Dr-Irv replied:
So I took your PR, and used this:
```python
    func: Callable[Concatenate[Self, P], TType]
        | tuple[Callable[Concatenate[Self, P], TType], _str],
```
and it passed the typing tests.

If that hadn't worked, then I would have just using an `overload`, where you specify the 2 possible values of `func`, but I don't think that's necessary.

paw-lu replied:
Thanks for giving it a test drive!

But reading your annotation, this means that a `func` tuple argument can only take in a function as the first element if the function takes in `Self`. This is not explicitly true for the `tuple` version (it is for the `Callable` version of the arg though!).

The `tuple` version should take in a function as the first element where any one parameter (not necessarily the first) takes in `Self`.

I don't think we have a test right now that would explicitly used .

It's not obvious to me how `overload` would fix this either to be honest :(.

## Example of what should be acceptable behavior

```
import pandas as pd
from pandas import DataFrame

def hi(x: int, y: DataFrame) -> DataFrame:
    return y

pd.DataFrame({"a": [1, 2, 3]}).pipe((hi, "y"), 1)
```

Dr-Irv replied:
> The `tuple` version should take in a function as the first element where any one parameter (not necessarily the first) takes in `Self`.

You're correct.  So I think what you have is correct (with the `...`), but tests should be added.

Thanks for working on this!

paw-lu commented on pandas-stubs/core/frame.pyi:
```diff
@@ -100,6 +103,7 @@ from pandas._typing import (
     MergeHow,
     NaPosition,
     NDFrameT,
+    P,
```
Yeah no problem!

Dr-Irv replied:
> Also, what's the rationale here (to be clear—not a pushback, just honestly curious)? I see both are used throughout the codebase.

Since we did an `import T as Ttype`, I figured it would be better to be consistent.

paw-lu replied:
These tests have now been implimented!

Dr-Irv replied:
> So this change has been undone—like you suggested—and moved to `generic.pyi` instead. There `TypeVar` is just specified as `T`—not `TType`—so I left `ParamSpec` as `P`—not `PType`. Happy to move to either in the end.

That's fine.

paw-lu commented:
>  I think if you took the concept proposed in this PR and used it in `generic.pyi` and removed it from `core`/`frame.pyi`
> So can you make that change, and then have tests for both `Series.pipe()` and `DataFrame.pipe()` ?

Got it, that sounds good to me!

paw-lu commented:
Just to try things out, implemented the more complicated function with positional-only args, keyword-only args, etc. This fails in current tests.

```pytb
tests/test_frame.py:1401: error: Expression is of type <nothing>, not "DataFrame"  [assert-type]
tests/test_frame.py:1403: error: Argument 1 to "pipe" of "DataFrame" has incompatible type "Callable[[DataFrame, int, list[float], str, NamedArg(tuple[int, int], 'keyword_only')], DataFrame]"; expected "Callable[[DataFrame, int, list[float], str, tuple[int, int]], <nothing>] | tuple[Callable[..., <nothing>], str]"  [arg-type]
Found 2 errors in 1 file (checked 224 source files)
```

However, when I upgrade mypy to latest version (1.7.1), the test does pass. However, it also raises a bunch of new errors elsewhere in code, I think all of them are of now unused `type: ignore` comments (easy fix).

<details>
<summary>mypy 1.7.1 errors</summary>

```
pandas-stubs/core/series.pyi:181: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/series.pyi:226: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/series.pyi:239: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/series.pyi:249: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/series.pyi:259: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/series.pyi:272: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/series.pyi:651: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/series.pyi:710: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/series.pyi:846: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/series.pyi:1179: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/series.pyi:1462: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/series.pyi:1496: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/series.pyi:1509: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/series.pyi:1523: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/series.pyi:1532: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/series.pyi:1556: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/series.pyi:1922: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/frame.pyi:280: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/frame.pyi:296: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/frame.pyi:304: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/frame.pyi:312: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/frame.pyi:320: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/frame.pyi:565: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/algorithms.pyi:28: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/algorithms.pyi:30: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/algorithms.pyi:32: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/reshape/tile.pyi:53: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/interval.pyi:215: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/interval.pyi:292: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/interval.pyi:294: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/interval.pyi:300: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/interval.pyi:302: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/interval.pyi:311: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/base.pyi:69: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/base.pyi:80: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/base.pyi:91: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/base.pyi:102: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/base.pyi:113: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/base.pyi:126: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/base.pyi:138: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/base.pyi:149: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/base.pyi:160: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/base.pyi:171: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/base.pyi:182: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/base.pyi:193: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/base.pyi:204: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/base.pyi:215: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/indexes/base.pyi:424: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/groupby/generic.pyi:148: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/core/groupby/generic.pyi:170: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/interval.pyi:202: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/interval.pyi:204: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/interval.pyi:206: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/interval.pyi:210: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/interval.pyi:212: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/interval.pyi:214: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/tslibs/timestamps.pyi:216: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/tslibs/timestamps.pyi:218: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/tslibs/timestamps.pyi:220: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/tslibs/timestamps.pyi:224: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/tslibs/timestamps.pyi:226: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/tslibs/timestamps.pyi:228: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/tslibs/timedeltas.pyi:311: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/tslibs/timedeltas.pyi:313: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/tslibs/timedeltas.pyi:315: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/tslibs/timedeltas.pyi:322: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/tslibs/timedeltas.pyi:324: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/tslibs/timedeltas.pyi:326: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/tslibs/period.pyi:102: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/tslibs/period.pyi:104: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/tslibs/period.pyi:106: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/tslibs/period.pyi:136: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/tslibs/period.pyi:138: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
pandas-stubs/_libs/tslibs/period.pyi:140: error: Unused "type: ignore" comment, use narrower [overload-overlap] instead of [misc] code  [unused-ignore]
```

</details>

So in order to implement this I'm thinking we need to:

1. Update mypy
2. Fix the new errors that come with this upgrade (which I think is just deleting a bunch of `# type: ignore`'s)

Happy to attempt this in another PR first if this is something we are interested in.

twoertwein commented:
Main uses mypy 1.7.1 now, so you can just rebase/merge and all the mypy warnings should disappear.

paw-lu commented:
Cool, rebased on top of the latest changes and things mostly seem to pass. Added a lot of tests for `frame.pyi`, they all seem to pass as expected now.

I have some local failures right now on:

<details>
<summary>1. pytest</summary>

```
========================================== short test summary info ===========================================
ERROR tests/test_api_types.py - DeprecationWarning: datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a fut...
ERROR tests/test_config.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_dtypes.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_errors.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_extension.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_frame.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_holidays.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_indexes.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_interval.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_interval_index.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_io.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_merge.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_pandas.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_plotting.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_resampler.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_scalars.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_series.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_styler.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_testing.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_timefuncs.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_utility.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
ERROR tests/test_windowing.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most like...
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 22 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================= 22 errors in 1.74s =============================================

===========================================
Step: 'Run pytest' failed!
===========================================
```
</details>

<details>
<summary>2. pyright</summary>

```
Poe => pyright
/Users/pawlu/Documents/personal/pandas-stubs/pandas-stubs/_libs/tslibs/timestamps.pyi
  /Users/pawlu/Documents/personal/pandas-stubs/pandas-stubs/_libs/tslibs/timestamps.pyi:103:9 - error: Method "fromtimestamp" overrides class "datetime" in an incompatible manner
    Parameter 2 name mismatch: base parameter is named "timestamp", override parameter is named "t" (reportIncompatibleMethodOverride)
1 error, 0 warnings, 0 informations
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/pawlu/Documents/personal/pandas-stubs/scripts/test/run.py", line 15, in pyright_src
    subprocess.run(cmd, check=True)
  File "/opt/homebrew/Cellar/python@3.12/3.12.1/Frameworks/Python.framework/Versions/3.12/lib/python3.12/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['pyright']' returned non-zero exit status 1.
```

</details>

I _think_ they're unrelated to these changes, but I'll switch back to main to check if they pass locally on my machine later before proceeding with the rest of the changes.

paw-lu commented:
Ok, seems that test that are current failing on pyright and pytest were failing in the commit prior to mine, so I'm going to ignore those. I've fully implemented tests for frame and implementation.

I plan to just do one more thing from here: add tests for pipe to `series.pyi`.

Dr-Irv commented on tests/test_frame.py:
```diff
@@ -1398,21 +1398,124 @@ def foo(df: pd.DataFrame) -> pd.DataFrame:
+                    dataframe_not_first_arg,
+                    "df",
+                ),
+                1,  # pyright: ignore[reportGeneralTypeIssues]
```
Created https://github.com/microsoft/pyright/issues/6796 for this.

Dr-Irv reviewed (changes requested):
This all looks pretty good.  Can you add tests in `test_series.py` as well similar to what you have for `test_frame.py` ?

paw-lu commented:
Added a tests in `test_series.py`. 

Now that it seems like

- https://github.com/microsoft/pyright/pull/6797

will make `overload` work better with pyright (thanks @Dr-Irv !), I'll probably move the implementation over to that.

Dr-Irv commented:
> Added a tests in `test_series.py`.
> 
> Now that it seems like
> 
> * [Fixed a bug in the function type assignment logic that leads to a fal… microsoft/pyright#6797](https://github.com/microsoft/pyright/pull/6797)
> 
> will make `overload` work better with pyright (thanks @Dr-Irv !), I'll probably move the implementation over to that.

Yes, so make the change in your PR to use the overloads (as in the example at https://github.com/microsoft/pyright/issues/6796#issuecomment-1868102290 ), and you might get a failure in the tests if you don't include an `ignore` comment, so we'll wait until the new pyright is released, bump up the pyright version, and then all will be well.

We probably will have to wait until 12/26 or later to get this all done.

Dr-Irv commented on tests/test_frame.py:
```diff
@@ -1398,21 +1398,125 @@ def foo(df: pd.DataFrame) -> pd.DataFrame:
+
+    check(
+        assert_type(
+            # FIXME: Remove ignore when pyright releases fix:
+            # https://github.com/microsoft/pyright/pull/6797
+            df.pipe(qux, 1, [1.0, 2.0], argument_2="hi", keyword_only=(1, 2)),  # pyright: ignore[reportGeneralTypeIssues]
```
pyright 1.1.343 is released.  Can you update `pyproject.toml` to have `pyright = ">=1.1.343` and then you can remove this?

Dr-Irv commented on tests/test_series.py:
```diff
@@ -2847,3 +2847,116 @@ def test_round() -> None:
+        assert_type(
+            # FIXME: Remove ignore when pyright releases fix:
+            # https://github.com/microsoft/pyright/pull/6797
+            ser.pipe(  # pyright: ignore[reportGeneralTypeIssues]
```
same comment as above about upgrading pyright

Dr-Irv reviewed (approved):
Thanks @paw-lu
